README.md
# ding-res-msg

[![Travis](https://img.shields.io/travis/yidinghan/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)
[![npm](https://img.shields.io/npm/l/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)
[![npm](https://img.shields.io/npm/v/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)
[![npm](https://img.shields.io/npm/dm/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)
[![David](https://img.shields.io/david/yidinghan/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)
[![David](https://img.shields.io/david/dev/yidinghan/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)
[![node](https://img.shields.io/node/v/ding-res-msg.svg?style=flat-square)](https://www.npmjs.com/package/ding-res-msg)

Response body formatter

<!-- TOC -->

-   [ding-res-msg](#ding-res-msg)
-   [Usage](#usage)
    -   [Installation](#installation)
    -   [Quick Start](#quick-start)
    -   [Examples](#examples)
-   [JSDoc](#jsdoc)
    -   [resMsg](#resmsg)

<!-- /TOC -->

# Usage

## Installation

```shell
npm i --save ding-res-msg
```

## Quick Start

```js
const resMsg = require('ding-res-msg');

// express
const controller = (req, res) => {
    const data = 'hello world';
    res.send(resMsg({ data }))
}

// koa2
const controller = (ctx, next) => {
    const data = 'hello world';
    ctx.body = resMsg({ data }); 
}
```

## Examples

```js
const resMsg = require('ding-res-msg');

console.log(resMsg());
// { success: true, data: undefined }

console.log(resMsg({
    data: {
        hello: 'world'
    }
}));
// { success: true, data: { hello: 'world' } }

console.log(resMsg({ error: new Error('test') }));
// { success: false, error: 'test', code: 400 }
```

More examples can be found on [jsdoc](#resmsg)

# JSDoc

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

## InputPayload

resMsg input arguments object define

Type: ([Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) \| [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object))

**Parameters**

-   `payload`   (optional, default `{}`)

**Properties**

-   `error` **([Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) \| [string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String))** failed response error
-   `data` **any** success response data
-   `code` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)?** failed response error code
-   `isPaging` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Whether to update the data object to msg
-   `isProduction` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Whether to add the stack to the msg,
                     if true will not add

## Message

resMsg return object define

Type: [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)

**Parameters**

-   `payload`   (optional, default `{}`)

**Properties**

-   `success` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** whether happend
-   `data` **any?** success response data
-   `code` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)?** failed response error code
-   `error` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** failed response error

## parseArguments

format input arguments

**Parameters**

-   `payload` **[InputPayload](#inputpayload)**  (optional, default `{}`)

Returns **[InputPayload](#inputpayload)** formated payload

## resMsg

res msg formattor

**Parameters**

-   `payload` **[InputPayload](#inputpayload)** input arguments or Error (optional, default `{}`)

**Examples**

```javascript
const resMsg = require('ding-res-msg');
console.log(resMsg());
// { success: true, data: undefined }

console.log(resMsg({ data: { total: 100 }, isPaging: true }));
// { success: true, total: 100 }

console.log(resMsg({ error: new Error('test') }));
// { success: false, error: 'test', code: 400 }

// Error field supports string error
console.log(resMsg({ error: 'test' }));
// { success: false, error: 'test', code: 400 }

// You can put the error directly in the first place
console.log(resMsg(new Error('test')));
// { success: false, error: 'test', code: 400 }

// Use error.code as msg.code
const error = new Error('test');
error.code = 503;
console.log(resMsg(error));
// { success: false, error: 'test', code: 503 }

// customised msg.code without error.code;
console.log(resMsg({ error: new Error('test'), code: 500 }));
// { success: false, error: 'test', code: 500 }

// NODE_ENV !== 'prod'
// You can get stack trace in the response body
// As long as you are not running in the production environment
console.log(resMsg(new Error('test')));
// { success: false, error: 'test', code: 400, stack: ['msg', '...'] }

// NODE_ENV !== 'prod'
// You cannot get stack trace in the response body
// event you are not running in a not prod environment
console.log(resMsg({ error: 'test', isProduction: true }));
// { success: false, error: 'test', code: 400 }

// NODE_ENV === 'prod'
// You can get stack trace in the response body
// event you are running in a prod environment
console.log(resMsg({ error: 'test', isProduction: false }));
// { success: false, error: 'test', code: 400, stack: ['msg', '...'] }

// use boom to create error
const boom = require('boom');
const error = boom.create(400)
console.log(resMsg(error));
// { success: false, error: 'Bad Request', code: 400, stack: ['Error', '...'] }
```

Returns **[Message](#message)** formatted response msg body,
                 if is failed msg and error have `code` or `statusCode`
                 msg.code would take that first


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.3.1"></a>
## [1.3.1](https://github.com/yidinghan/ding-res-msg/compare/v1.3.0...v1.3.1) (2017-10-30)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/yidinghan/ding-res-msg/compare/v1.2.1...v1.3.0) (2017-07-26)


### Features

* **error:** support boom error ([6f198d4](https://github.com/yidinghan/ding-res-msg/commit/6f198d4))



<a name="1.2.1"></a>
## [1.2.1](https://github.com/yidinghan/ding-res-msg/compare/v1.2.0...v1.2.1) (2017-06-05)



<a name="1.2.0"></a>
# [1.2.0](https://github.com/yidinghan/ding-res-msg/compare/v1.1.0...v1.2.0) (2017-06-05)


### Features

* **stack:** support customized isProduction flag ([00f6e56](https://github.com/yidinghan/ding-res-msg/commit/00f6e56))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/yidinghan/ding-res-msg/compare/v1.0.0...v1.1.0) (2017-05-15)


### Features

* **res msg:** use error itself when not error.message found ([d7997bb](https://github.com/yidinghan/ding-res-msg/commit/d7997bb))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/yidinghan/ding-res-msg/compare/v0.2.2...v1.0.0) (2017-05-15)


### Features

* **res msg:** only use error.message as msg.error ([d79a598](https://github.com/yidinghan/ding-res-msg/commit/d79a598))
* **res msg:** rename isPagination to isPaging ([67f59d0](https://github.com/yidinghan/ding-res-msg/commit/67f59d0))



<a name="0.2.2"></a>
## [0.2.2](https://github.com/yidinghan/ding-res-msg/compare/v0.2.1...v0.2.2) (2017-05-15)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/yidinghan/ding-res-msg/compare/v0.2.0...v0.2.1) (2017-05-15)



<a name="0.2.0"></a>
# 0.2.0 (2017-05-15)


### Features

* init res msg file ([ecbae54](https://github.com/yidinghan/ding-res-msg/commit/ecbae54))


test/index.test.js
const { test } = require('ava');
const boom = require('boom');

const resMsg = require('../');


const has = Object.hasOwnProperty;

test('should success handler boom error 503', (t) => {
  const msg = resMsg(boom.create(503));

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'Service Unavailable', 'default boom 503 error message');
  t.is(code, 503);
});

test('should success handler boom error 400', (t) => {
  const msg = resMsg(boom.create(400));

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'Bad Request', 'default 400 boom error');
  t.is(code, 400);
});

test('get success msg without any input', (t) => {
  const msg = resMsg();

  t.is(Object.keys(msg).length, 2, 'should only have 2 keys');
  t.true(has.call(msg, 'success'));
  t.true(has.call(msg, 'data'));

  const { success, data } = msg;
  t.true(success, 'success flag');
  t.is(data, undefined, 'should be undefined when not giving data');
});

test('get failed msg with first arg is Error', (t) => {
  const msg = resMsg(new Error('test'));

  t.is(Object.keys(msg).length, 4, 'should only have 4 keys');
  t.true(has.call(msg, 'success'));
  t.true(has.call(msg, 'error'));
  t.true(has.call(msg, 'code'));
  t.true(has.call(msg, 'stack'));

  const { success, error, code, stack } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 400, 'should use default code');
  t.true(Array.isArray(stack), 'should be splited to array');
  t.is(stack[0], 'Error: test');
});

test('get failed msg with object input', (t) => {
  const msg = resMsg({ error: new Error('test') });

  const { success, error, code, stack } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 400, 'should use default code');
  t.true(Array.isArray(stack), 'should be splited to array');
});

test('get failed msg with object input', (t) => {
  const msg = resMsg({ error: new Error('test'), code: 500 });

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 500, 'should use customised code');
});

test('get failed msg with msg.code is error.code', (t) => {
  const err = new Error('test');
  // @ts-ignore
  err.code = 401;
  const msg = resMsg({ error: err });

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 401);
});

test('get failed msg with msg.statusCode is error.code', (t) => {
  const err = new Error('test');
  // @ts-ignore
  err.statusCode = 403;
  const msg = resMsg({ error: err });

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 403);
});

test('Get the failed msg by customizing error.stack', (t) => {
  const err = new Error('test');
  // @ts-ignore
  err.stack = { raw: err.stack, stack: err.stack.split('\n') };
  const msg = JSON.parse(JSON.stringify(resMsg({ error: err })));

  const { success, error, code, stack } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 400);
  t.deepEqual(stack, err.stack);
});

test('get failed msg without stack when isProduction:false', (t) => {
  const msg = resMsg({ error: 'test', isProduction: true });

  t.false(has.call(msg, 'stack'));
});

test('get failed msg with array object', (t) => {
  const msg = resMsg({ error: ['test'] });

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.deepEqual(error, ['test']);
  t.is(code, 400);
});

test('get failed msg with string error in payload field', (t) => {
  const msg = resMsg({ error: 'test' });

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 400);
});

test('get success msg with data', (t) => {
  const msg = resMsg({ data: 'success' });

  const { success, data } = msg;
  t.true(success, 'success flag');
  t.is(data, 'success');
});

test('get success msg with isPagination:true', (t) => {
  const msg = resMsg({ data: { total: 100 }, isPaging: true });

  const { success, data, total } = msg;
  t.true(success, 'success flag');
  t.is(data, undefined);
  t.is(total, 100);
});


test/index.prod.test.js
const { test } = require('ava');

process.env.NODE_ENV = 'prod';

const resMsg = require('../');

const has = Object.hasOwnProperty;

test('get failed msg without stack when NODE_ENV:prod', (t) => {
  const msg = resMsg(new Error('test'));

  t.is(Object.keys(msg).length, 3, 'should only have 3 keys');
  t.true(has.call(msg, 'success'));
  t.true(has.call(msg, 'error'));
  t.true(has.call(msg, 'code'));
  t.false(has.call(msg, 'stack'));

  const { success, error, code } = msg;
  t.false(success, 'success flag');
  t.is(error, 'test', 'should use error.message');
  t.is(code, 400, 'should use default code');
});

test('get failed msg with stack when NODE_ENV:prod but isProduction:true', (t) => {
  const msg = resMsg({ error: 'test', isProduction: false });

  t.is(Object.keys(msg).length, 4, 'should only have 4 keys');
  t.true(has.call(msg, 'stack'));

  const { stack } = msg;
  t.true(Array.isArray(stack), 'should be splited to array');
  t.is(stack[0], 'Error: test');
});


index.js
const isProd = /prod/.test(process.env.NODE_ENV);

/**
 * resMsg input arguments object define
 * @typedef {Error|Object} InputPayload
 * @property {Error|string} error - failed response error
 * @property {*} data - success response data
 * @property {number} [code=400] - failed response error code
 * @property {boolean} [isPaging=false] - Whether to update the data object to msg
 * @property {boolean} [isProduction] - Whether to add the stack to the msg,
 *                  if true will not add
 */

/**
 * resMsg return object define
 * @typedef {Object} Message
 * @property {boolean} [success=true] - whether happend
 * @property {*} [data] - success response data
 * @property {number} [code=400] - failed response error code
 * @property {string} [error] - failed response error
 */

/**
 * format input arguments
 *
 * @param {InputPayload} payload
 * @return {InputPayload} formated payload
 */
const parseArguments = (payload = {}) => {
  if (payload.isBoom) {
    return {
      error: payload,
      code: payload.output.statusCode,
    };
  } else if (payload instanceof Error) {
    return { error: payload };
  }

  const error = payload.error;
  if (error !== undefined && typeof error === 'string') {
    return Object.assign(payload, {
      error: new Error(error),
    });
  }

  return payload;
};

/**
 * res msg formattor
 *
 * @param {InputPayload} [payload={}] - input arguments or Error
 * @return {Message} formatted response msg body,
 *                  if is failed msg and error have `code` or `statusCode`
 *                  msg.code would take that first
 * @example
 * const resMsg = require('ding-res-msg');
 * console.log(resMsg());
 * // { success: true, data: undefined }
 *
 * console.log(resMsg({ data: { total: 100 }, isPaging: true }));
 * // { success: true, total: 100 }
 *
 * console.log(resMsg({ error: new Error('test') }));
 * // { success: false, error: 'test', code: 400 }
 *
 * // Error field supports string error
 * console.log(resMsg({ error: 'test' }));
 * // { success: false, error: 'test', code: 400 }
 *
 * // You can put the error directly in the first place
 * console.log(resMsg(new Error('test')));
 * // { success: false, error: 'test', code: 400 }
 *
 * // Use error.code as msg.code
 * const error = new Error('test');
 * error.code = 503;
 * console.log(resMsg(error));
 * // { success: false, error: 'test', code: 503 }
 *
 * // customised msg.code without error.code;
 * console.log(resMsg({ error: new Error('test'), code: 500 }));
 * // { success: false, error: 'test', code: 500 }
 *
 * // NODE_ENV !== 'prod'
 * // You can get stack trace in the response body
 * // As long as you are not running in the production environment
 * console.log(resMsg(new Error('test')));
 * // { success: false, error: 'test', code: 400, stack: ['msg', '...'] }
 *
 * // NODE_ENV !== 'prod'
 * // You cannot get stack trace in the response body
 * // event you are not running in a not prod environment
 * console.log(resMsg({ error: 'test', isProduction: true }));
 * // { success: false, error: 'test', code: 400 }
 *
 * // NODE_ENV === 'prod'
 * // You can get stack trace in the response body
 * // event you are running in a prod environment
 * console.log(resMsg({ error: 'test', isProduction: false }));
 * // { success: false, error: 'test', code: 400, stack: ['msg', '...'] }
 *
 * // use boom to create error
 * const boom = require('boom');
 * const error = boom.create(400)
 * console.log(resMsg(error));
 * // { success: false, error: 'Bad Request', code: 400, stack: ['Error', '...'] }
 */
const resMsg = (payload) => {
  const {
    error,
    data,
    code = 400,
    isPaging = false,
    isProduction,
  } = parseArguments(payload);

  let msg = { success: true, data };
  if (!error) {
    if (isPaging) {
      delete msg.data;
      Object.assign(msg, data);
    }
    return msg;
  }

  const finalCode = error.code || error.statusCode || code;
  // @ts-ignore
  msg = {
    success: false,
    error: error.message || error,
    code: finalCode,
  };

  const isNotStack =
    (isProduction === undefined ? isProd : isProduction) === true;
  if (isNotStack === false) {
    msg.stack =
      typeof error.stack === 'string' ? error.stack.split('\n') : error.stack;
  }

  return msg;
};

module.exports = resMsg;
